src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -9,295 +9,445 @@
+            pslr.LowerDecisionTree(expression, node.DecisionTree, result);
+
+            // if the endpoint is reachable, we exit the switch
+            if (!node.DecisionTree.MatchIsComplete) result.Add(_factory.Goto(node.BreakLabel));
+            // at this point the end of result is unreachable.
 
-            // output the sections of code (that were reachable)
+            // output the sections of code
             foreach (var section in node.SwitchSections)
             {
-                ArrayBuilder<BoundStatement> sectionBuilder = null;
+                // Start with the part of the decision tree that is in scope of the section variables.
+                // Its endpoint is not reachable (it jumps back into the decision tree code).
+                var sectionBuilder = pslr.SwitchSections[section];
+
+                // Add labels corresponding to the labels of the switch section.
                 foreach (var label in section.SwitchLabels)
                 {
-                    if (usedLabels.Contains(label.Label))
-                    {
-                        if (sectionBuilder == null)
-                        {
-                            sectionBuilder = ArrayBuilder<BoundStatement>.GetInstance();
-                        }
-
-                        sectionBuilder.Add(_factory.Label(label.Label));
-                    }
+                    sectionBuilder.Add(_factory.Label(label.Label));
                 }
 
-                if (sectionBuilder != null)
-                {
-                    sectionBuilder.AddRange(VisitList(section.Statements));
-                    sectionBuilder.Add(_factory.Goto(node.BreakLabel));
-                    result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
-                }
+                // Add the translated body of the switch section
+                sectionBuilder.AddRange(VisitList(section.Statements));
+                sectionBuilder.Add(_factory.Goto(node.BreakLabel));
+                result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
+                // at this point the end of result is unreachable.
             }
 
             result.Add(_factory.Label(node.BreakLabel));
-            return _factory.Block(usedTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            var translatedSwitch = _factory.Block(pslr.DeclaredTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            return translatedSwitch;
         }
 
-        private void LowerDecisionTree(DecisionTree decisionTree, HashSet<LabelSymbol> usedLabels, HashSet<LocalSymbol> usedTemps, ArrayBuilder<BoundStatement> result)
+        private class PatternSwitchLocalRewriter
         {
-            if (decisionTree == null) return;
-            switch (decisionTree.Kind)
+            public readonly LocalRewriter LocalRewriter;
```

Comment:
Since the enclosing class is private, they remain private to its enclosing type.

---
In reply to: [70326450](https://github.com/dotnet/roslyn/pull/12416#discussion_r70326450) [](ancestors = 70326450)